### PR TITLE
Zingo sync add client module

### PR DESCRIPTION
**Pull request by Oscar-Pepper**
this approach will have the efficiency gains of the client being handled by a single task but will allow multiple other tasks to make requests concurrently. this also allows request prioritisation so small data fetches are not blocked on queued larger data fetches to complete and we can compare sync performance to varied prioritisation schemes.

### CODE REVIEW

File: zingo-sync/src/client/fetcher.rs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+/// prioritisation and fetching from the server.
+/// Returns the data specified in the [`self::FetchRequest`] variant via the provided [`tokio::sync::mpsc::oneshot::Sender`].
+///
+/// Allows all requests to the server to be handled from a single task for efficiency and also enables
+/// request prioritisation for further performance enhancement
+pub async fn fetcher(
+    mut fetch_request_receiver: UnboundedReceiver<FetchRequest>,
+    mut client: CompactTxStreamerClient<zingo_netutils::UnderlyingService>,
+) -> Result<(), ()> {
+    let mut fetch_request_queue: Vec<FetchRequest> = Vec::new();
```
Comment (idky137) [inline zingo-sync/src/client/fetcher.rs:27]: Have you thought about using a mcsp::channel directly as the queue? It acts as a fifo list by default so actually works quite well as a queue. Sender threads push directly onto the channel and the fetcher pops them. I think there is some added complexity using a vec as a fifo list (better for filo)..

Comment (Oscar-Pepper) [reply zingo-sync/src/client/fetcher.rs:27]: good point, i think we can use a more optimal queue for this reason but it is not designed to be FIFO, we want to prioritise requests. I think your idea about Ordering may be a good solution but as the backpressure should be very low I dont think we need to optimise this yet

File: zingo-netutils/src/lib.rs [hunk 2]
```diff
@@ -16,7 +16,8 @@ use tonic::Status;
 use tower::util::BoxCloneService;
 use zcash_client_backend::proto::service::compact_tx_streamer_client::CompactTxStreamerClient;
 
-type UnderlyingService = BoxCloneService<
+/// TODO: add doc-comment
```
Comment (zancas) [inline zingo-netutils/src/lib.rs:19]: Why is this type being defined?

File: zingo-sync/src/client.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
+//! Module for handling all connections to the server
+
+use std::ops::Range;
+
+use zcash_client_backend::proto::{compact_formats::CompactBlock, service::BlockId};
```
Comment (zancas) [inline zingo-sync/src/client.rs:5]: It's important to double check with zingolib to see if there's another definition of `CompactBlock` dangling around.

Comment (Oscar-Pepper) [reply zingo-sync/src/client.rs:5]: yeah good idea. i have checked and this is the same one used in zingolib currently and in LRZ etc. as defined in the proto files

File: zingo-netutils/src/lib.rs (hunk 2 above)
Comment (Oscar-Pepper) [reply zingo-netutils/src/lib.rs:19]: well, you may have a better idea here but I think it's a lot to define without this type alias. any suggestions welcome

**Comment by zancas** [comment]
> this approach will have the efficiency gains of the client being handled by a single task

What efficiency gain is this?

File: zingo-netutils/src/lib.rs (hunk 2 above)
Comment (zancas) [reply zingo-netutils/src/lib.rs:19]: Please replace the TODO: with something else.

Comment (Oscar-Pepper) [reply zingo-netutils/src/lib.rs:19]: Yeah sure

**Comment by Oscar-Pepper** [comment]
> > this approach will have the efficiency gains of the client being handled by a single task
> 
> What efficiency gain is this?

Me and @idky137 conducted a test on codeshare where two tasks were requesting blocks simultaneously and compared this to two tasks where only one was requesting the same total number of blocks. When only one task was connecting to the server the blocks came in faster

File: zingo-netutils/src/lib.rs (hunk 2 above)
Comment (zancas) [reply zingo-netutils/src/lib.rs:19]: Oh I see it's just being made `pub`...

File: zingo-sync/src/client.rs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+) -> Result<BlockHeight, ()> {
+    let (sender, receiver) = oneshot::channel::<BlockId>();
+    fetch_request_sender
+        .send(FetchRequest::ChainTip(sender))
+        .unwrap();
+    let chain_tip = receiver.await.unwrap();
+
+    Ok(BlockHeight::from_u32(chain_tip.height as u32))
+}
+/// Gets the height of the blockchain from the server.
```
Comment (zancas) [inline zingo-sync/src/client.rs:34]: This doc-comment seems out of date... or associated with the wrong fn.

Comment (Oscar-Pepper) [reply zingo-sync/src/client.rs:34]: Yeah I noticed that, it's fixed in the following PR

**Review by zancas — APPROVED** [review]

**Pull request merged**